src/vs/workbench/api/common/extHostLanguageModels.ts

```diff
@@ -291,54 +291,62 @@ export class ExtHostLanguageModels implements ExtHostLanguageModelsShape {
 		this._onDidChangeProviders.fire(undefined);
 	}
 
+	async getLanguageModelByIdentifier(extension: IExtensionDescription, identifier: string): Promise<vscode.LanguageModelChat | undefined> {
+
+		const data = this._allLanguageModelData.get(identifier);
+		if (!data) {
+			// model gone? is this an error on us?
+			return;
+		}
+
+		// make sure auth information is correct
+		if (this._isUsingAuth(extension.identifier, data.metadata)) {
+			await this._fakeAuthPopulate(data.metadata);
+		}
+
+		let apiObject = data.apiObjects.get(extension.identifier);
+		if (!apiObject) {
+			const that = this;
+			apiObject = {
+				id: data.metadata.id,
+				vendor: data.metadata.vendor,
+				family: data.metadata.family,
+				version: data.metadata.version,
+				name: data.metadata.name,
+				maxInputTokens: data.metadata.maxInputTokens,
+				countTokens(text, token) {
+					if (!that._allLanguageModelData.has(identifier)) {
+						throw extHostTypes.LanguageModelError.NotFound(identifier);
+					}
+					return that._computeTokenLength(identifier, text, token ?? CancellationToken.None);
+				},
+				sendRequest(messages, options, token) {
+					if (!that._allLanguageModelData.has(identifier)) {
+						throw extHostTypes.LanguageModelError.NotFound(identifier);
+					}
+					return that._sendChatRequest(extension, identifier, messages, options ?? {}, token ?? CancellationToken.None);
+				}
+			};
+
+			Object.freeze(apiObject);
+			data.apiObjects.set(extension.identifier, apiObject);
+		}
+
+		return apiObject;
+	}
+
 	async selectLanguageModels(extension: IExtensionDescription, selector: vscode.LanguageModelChatSelector) {
 
 		// this triggers extension activation
 		const models = await this._proxy.$selectChatModels({ ...selector, extension: extension.identifier });
```

Comment:
This reminds me that we might be missing an activation event, in theory at least. Assume the chat model is from an extension that isn't activated yet but picked in the UI. In that case, we might need some client-side call to trigger the chat-activation event, like this https://github.com/microsoft/vscode/blob/46c3cc46a159f9140637ef86de34b27144ee09e3/src/vs/workbench/contrib/chat/common/languageModels.ts#L246

Reply:
Oh right

Reply:
I'm not sure where in the flow that should happen. We would have to activate all the providers as soon as the picker is shown right? Which undoes some of the work to make things nicely lazy. I'll leave it alone for now.